lib/sshkit/backends/netssh/sftp_transfer.rb

```diff
@@ -0,0 +1,46 @@
+require "net/sftp"
+
+module SSHKit
+  module Backend
+    class Netssh < Abstract
+      class SftpTransfer
+        def initialize(ssh, summarizer)
+          @ssh = ssh
+          @summarizer = summarizer
+        end
+
+        def upload!(local, remote, options)
+          options = { progress: self }.merge(options || {})
+          ssh.sftp.connect!
+          ssh.sftp.upload!(local, remote, options)
+        ensure
+          ssh.sftp.close_channel
+        end
+
+        def download!(remote, local, options)
+          options = { progress: self }.merge(options || {})
+          destination = local ? local : StringIO.new.tap { |io| io.set_encoding('BINARY') }
+
+          ssh.sftp.connect!
+          ssh.sftp.download!(remote, destination, options)
+          local ? true : destination.string
+        ensure
+          ssh.sftp.close_channel
+        end
+
+        def on_get(download, entry, offset, data)
+          entry.size ||= download.sftp.file.open(entry.remote, &:size)
+          summarizer.call(nil, entry.remote, offset + data.bytesize, entry.size)
+        end
```

Comment:
🗒️ net-sftp calls `on_get` to provide download progress. This information is much different that how net-scp provides progress. In net-sftp, `offset` refers to the offset of the local destination, _before_ the bytes have been transferred. So some arithmetic is needed to calculate the progress: `offset + data.bytesize`.

Furthermore, the total size of the file being transferred, which is supposed to be provided via `entry.size`, is `nil`. To work around this, I had to explicitly make an sftp `open` call to read the file size.